caddyhttp/staticfiles/fileserver.go

```diff
@@ -121,13 +125,41 @@ func (fs FileServer) serveFile(w http.ResponseWriter, r *http.Request, name stri
 		return http.StatusNotFound, nil
 	}
 
+	filename := d.Name()
+
+	for _, encoding := range staticEncodingPriority {
+		if strings.Contains(r.Header.Get("Accept-Encoding"), encoding) {
+			gf, err := fs.Root.Open(location + staticEncoding[encoding])
+
+			if err == nil {
+				stat, err := gf.Stat()
+
+				if err == nil {
+					// Close previous file - release fd
+					f.Close()
+
+					d = stat
+					f = gf
+
+					w.Header().Add("Vary", "Accept-Encoding")
+					w.Header().Set("Content-Encoding", encoding)
+
+					defer f.Close()
+					break
+				}
+
+				gf.Close()
```

Comment:
This loop is constructed a bit "deep" in my opinion. Can we do this instead, I find this much easier to follow (untested, please try it out if you like it!):

```go
// see if there is a compressed version of the file available; if so, use it
for _, encoding := range staticEncodingPriority {
	if !strings.Contains(r.Header.Get("Accept-Encoding"), encoding) {
		continue
	}
	encodedFile, err := fs.Root.Open(location + staticEncoding[encoding])
	if err != nil {
		encodedFile.Close()
		continue
	}
	encodedFileInfo, err := encodedFile.Stat()
	if err != nil {
		encodedFile.Close()
		continue
	}
	
	f.Close() // release fd of previously-selected file
	f = encodedFile
	d = encodedFileInfo

	w.Header().Add("Vary", "Accept-Encoding")
	w.Header().Set("Content-Encoding", encoding)

	defer f.Close()
	break
}
```